### PR TITLE
Add sstable disk size metric [work in progress]

### PR DESCRIPTION
**Pull request by lersek**
Add sstable disk size metric [work in progress]

### CODE REVIEW

**Comment by lersek** [comment]
@raphaelsc this is my first, super naïve and known-broken patch for #18191. At this time it should not be merged by any means (well the test suite fails anyway, so it couldn't be). I'm only opening a PR only for lack of a more suitable medium, for asking questions against actual code.

The patch crashes scylla basically immediately upon startup; the `bytes_on_disk()` call made from `sstable::seal_sstable` aborts, due to `_data_file_size` being zero.

If I remove that call from `sstable::seal_sstable`, then scylla no longer crashes upon startup, but it still crashes in several unit tests. These crashes occur for the same reason in `bytes_on_disk()` (i.e., for `_data_file_size` not having been set). The culprit call is most likely the one in `sstable::unlink`.

The calls in both `sstable::load` variants should be fine. As much as I could prove, those calls are made like this:

~~~
load (test variant)
  load_metadata
    read_toc
      do_read_simple
        increase _metadata_size_on_disk
  open_data
    update_info_for_opened_data
      set _data_file_size
      set _index_file_size
  bytes_on_disk()

load (main variant)
  set _metadata_size_on_disk
  update_info_for_opened_data
    set _data_file_size
    set _index_file_size
  bytes_on_disk()
~~~

i.e., when the `load` variants call `bytes_on_disk()`, all three members that `bytes_on_disk()` depends upon will have been set.

Is there a *comprehesnive sstable class design document* / state diagram that describes the allowed member function call orders / state transitions?

I'm unable to identify such locations that satisfy both of the following conditions:
- `bytes_on_disk()` is safe to call (i.e., all three pre-requisite sizes have been computed),
- updating the "sstables' disk occupation" metric is appropriate -- that is, when a previously written sstable is "discovered" (?), or when a new one is written / sealed (?), or when an sstable is removed (after compaction). In particular, "discovery" may not mean "loading", I presume.

It's possible that `bytes_on_disk()`, in its current form, is simply unusable for the new metric, exactly because it depends on such state transitions that simply don't occur early enough for refreshing the metric. In other words, we might have to calculate from scratch the metadata size, the data file size, and the index file size -- all of which seems like a bunch of otherwise unjustified *async* work (i.e., just for setting the metric upon startup). What do you think about that? Thanks.

File: sstables/sstables.cc [hunk 1]
```diff
@@ -1603,6 +1619,9 @@ future<> sstable::load(sstables::foreign_sstable_open_info info) noexcept {
         validate_partitioner();
         return update_info_for_opened_data();
     }).then([this] {
+        return bytes_on_disk_uncached();
+    }).then([this] (uint64_t bytes) {
```
Comment (raphaelsc) [inline sstables/sstables.cc:1623]: I don't think we need this. At this point, bytes_on_disk() can return correct result since _metadata_size_on_disk + _data_file_size + _index_file_size are all set.

Comment (raphaelsc) [reply sstables/sstables.cc:1623]: "I can't use the existent bytes_on_disk() function for this purpose.
Namely, since at least commit https://github.com/scylladb/scylladb/commit/2dbae856f84ebef8267e53ee542cc9aa1da27775 ("sstable: Piggyback on sstable
parser and writer to provide bytes_on_disk", 2023-04-27), bytes_on_disk()
insists on "_data_file_size" and "_index_file_size" having been set (to
nonzero values). However, at least "_data_file_size" may not be set when
we call bytes_on_disk() from seal_sstable(), for example."

I think we can do it in open_data(). see that _manager.increment_total_reclaimable_memory_and_maybe_reclaim() happens there, once the sstable becomes ready for reads.

File: sstables/sstables.cc [hunk 2]
```diff
@@ -1588,6 +1603,7 @@ future<> sstable::load(const dht::sharder& sharder, sstable_open_config cfg) noe
                 std::vector<unsigned>{this_shard_id()} : compute_shards_for_this_sstable(sharder);
     }
     co_await open_data(cfg);
+    _stats.on_seal_or_load(co_await bytes_on_disk_uncached());
 }
```
Comment (raphaelsc) [inline sstables/sstables.cc:1607]: I think we don't want to bump it here, since a sstable can be loaded at one shard, and will be transferred to another shard during boot (without unlink).

File: sstables/sstables.cc (hunk 1 above)
Comment (lersek) [reply sstables/sstables.cc:1623]: @raphaelsc please see my comment [here](https://github.com/scylladb/scylladb/pull/20574#issuecomment-2347099183). `bytes_on_disk` is indeed fine for both `load` variants; that was what I tried first. `open_data` is listed there, as a participant in one of the `load` call trees; `open_data` populates `_data_file_size` and `_index_file_size` for that `load` variant. However:
1. You are [suggesting](https://github.com/scylladb/scylladb/pull/20574#discussion_r1761809170) not to modify the `load` variant that calls `open_data`. This may be a valid suggestion for that particular `load` variant, but then the recommendation about `open_data` seems confusing (it's a contradiction), because `open_data` is called from exactly that `load` variant, so the accounting would still happen there.
2. Furthermore, `open_data` is not called from any other location (at least inside the class). It is a public function though, and seems to be called from outside the class. All in all I can't determine what `open_data`'s official purpose is, and what general impact it would have if we added to accounting to `open_data`. What we need is a central (unique) entry point that stands for "discovery", but I'm not sure which sst member function that is.
3. Another objection (or question) is that `open_data` is never called on the write path. So we wouldn't account for disk usage right after a new sst is written, until it gets loaded (which may not happen for an indefinite time, IIUC). From your [original comment](https://github.com/scylladb/scylladb/issues/18191#issuecomment-2049050978), it seemed that `seal_sstable` was a good hook point on the write path. However, when I [tried to call](https://github.com/scylladb/scylladb/pull/20574#issuecomment-2347099183) `bytes_on_disk` from `seal_sstable`, it crashed, due to `_data_file_size` not having been set. So where do I set the index and data file sizes on the write path? (Again, `open_data` is not a part of that path.)
4. The above bullets are all about *incrementing* the stat (on the read path, and on the write path). However, we also need to decrement the stat; `unlink` seems like the place to do it. However, `unlink` can happen on the write path too (`ctor -> open_sstable -> seal_sstable -> close_files -> unlink`). And just as in `seal_sstable`, `_data_file_size` is unavailable in `unlink` on the write path, so I can't call `bytes_on_disk` from `unlink` on the write path.

File: sstables/sstables.cc [hunk 3]
```diff
@@ -1601,10 +1601,10 @@ future<> sstable::load(sstables::foreign_sstable_open_info info) noexcept {
...
         validate_max_local_deletion_time();
         validate_partitioner();
-        return update_info_for_opened_data().then([this]() {
-            _total_reclaimable_memory.reset();
-            _manager.increment_total_reclaimable_memory_and_maybe_reclaim(this);
-        });
+        return update_info_for_opened_data();
+    }).then([this] {
+        _total_reclaimable_memory.reset();
+        _manager.increment_total_reclaimable_memory_and_maybe_reclaim(this);
```
Comment (avikivity) [inline sstables/sstables.cc:1626]: Better to just convert to a coroutine.

File: sstables/sstables.cc [hunk 4]
```diff
@@ -1559,6 +1559,21 @@ future<> sstable::reload_reclaimed_components() {
     sstlog.info("Reloaded bloom filter of {}", get_filename());
 }
 
+future<uint64_t> sstable::bytes_on_disk_uncached() const noexcept {
+    uint64_t bytes = 0;
+
+    co_await coroutine::parallel_for_each(_recognized_components, [this, &bytes] (const component_type& type) -> future<> {
+        try {
+            file f = co_await _storage->open_component(*this, type, open_flags::ro, file_open_options{}, false);
+            struct stat st = co_await f.stat().finally([&f] { return f.close(); });
```
Comment (avikivity) [inline sstables/sstables.cc:1568]: Wrinkles nose at coroutine/continuation mixes.

Is this really necessary? Can't we collect the size when we open the files (or when we append to them/delete them) rather that in a separate procedure?

File: sstables/sstables.cc [hunk 5]
```diff
@@ -3191,6 +3212,9 @@ future<> init_metrics() {
         sm::make_counter("total_deleted", [] { return sstables_stats::get_shard_stats().deleted; },
             sm::description("Counter of deleted sstables")),
 
+        sm::make_gauge("bytes_on_disk", [] { return sstables_stats::get_shard_stats().bytes_on_disk; },
+            sm::description("Bytes currently used on disk by sstables")),
+
```
Comment (avikivity) [inline sstables/sstables.cc:3217]: We could consider making this part of sstables_manager, or adding a label to support different storage classes. Could be added later. /cc @xemul

File: sstables/sstables.cc (hunk 4 above)
Comment (lersek) [reply sstables/sstables.cc:1568]: > Wrinkles nose at coroutine/continuation mixes.

Justifiedly so. I had four options (as far as I could tell):
- Don't close `f` explicitly, just let the destructor (?) deal with it. My own wrinkle factor: 3.
- Close `f` explicitly, using implementing the "finally" logic with explicit try / catch / std::current_exception. Wrote the code and cringed hard. Wrinkle factor: 3+.
- Browsed the source code and found prior art in (e.g.) `sstables/storage.cc`: `co_await with_closeable`. Generic, but hard to read. Wrinkle factor: 2.
- Figured that I might as well open-code `with_closeable` by chaining `.finally`, for such a short lambda. Wrinkle factor: 1 (not zero, but least annoying).

I can do `finally` with explicit try/catch/etc, but it was eye-poppingly awkard, for me. Please confirm if you want me to do that (or, of course, if I'm missing the obvious idiomatic solution, please let me know that one).

> Is this really necessary? Can't we collect the size when we open the files (or when we append to them/delete them) rather that in a separate procedure?

This is the question I've been asking since first looking at #18191. One of the first things I asked for was a design document or some kind of state diagram that described the state transitions / calling order of APIs. If we know what happens after what, then we can set up new data dependencies, i.e., precompute the disk usage and then just return that value. Thus far nobody shared any insight about the state transitions; so I'm getting the impression that the APIs and their interdependencies are more or less ad-hoc.

Metadata size may already be calculated in advance, but I don't understand what (and when) decides whether the sst will have an index file and/or a data file. In particular, if you look at commit 2dbae856f84e, `bytes_on_disk` was changed to require all three to be nonzero (metadata, index, data). This requirement is not satisfied however when scylla starts up with an otherwise empty database, and compacts some system-level sstables. Commit 2dbae856f84e didn't explain why requiring all three kinds of data for disk size calculation was justified.

In summary, I calculated the disk usage like this in v1 _precisely_ in order to force the question out in the open. The proposed "always re-stat" approach sucks, but I don't see the alternative, as long as we can't exactly specify the state transitions / API call order.

(Sorry about the wall of text.)

File: sstables/sstables.cc (hunk 5 above)
Comment (lersek) [reply sstables/sstables.cc:3217]: > We could consider making this part of sstables_manager

@raphaelsc indicated something similar in the original bug report #18191, but (a) I found stats being set up in a bunch of -- apparently -- unrelated locations and functions already, and (b) I figured if it could be kept within the `sstable` class, then why would we want to drag `sstables_manager` into it? Isolation of responsibilites etc.

> or adding a label to support different storage classes

I don't understand this. What are labels? Thanks.

(I assume by "storage classes" you mean filesystem storage vs S3 storage.)

File: sstables/sstables.cc (hunk 2 above)
Comment (lersek) [reply sstables/sstables.cc:1607]: Unfortunately, if I remove this call, then the stat underflows when unlinking subtracts the sizes of the files being removed.

If I keep this call, then after scylla startup, `http://localhost:9180/metrics` reports
~~~
# HELP scylla_sstables_bytes_on_disk Bytes currently used on disk by sstables
# TYPE scylla_sstables_bytes_on_disk gauge
scylla_sstables_bytes_on_disk{shard="0"} 455755.000000
~~~
and 
~~~
find $SCYLLA_HOME/data/ -type f -printf '%s\n' | (X=0; while read Y; do X=$((X+Y)); done; echo $X)
~~~
also prints 455755 precisely.

File: sstables/sstables.cc (hunk 1 above)
Comment (lersek) [reply sstables/sstables.cc:1623]: > I don't think we need this. At this point, bytes_on_disk() can return correct result since _metadata_size_on_disk + _data_file_size + _index_file_size are all set.
> I think we can do it in open_data().

Doesn't work; if I  modify both load variants to use bytes_on_disk(), boost/sstable_test fails (big_summary_query_0, big_summary_query_32, all_in_place).

File: sstables/sstables.cc (hunk 4 above)
Comment (lersek) [reply sstables/sstables.cc:1568]: I'll use `co_await with_closeable` in v2.

**Comment by lersek** [comment]
v2:
- patch#1: rather than unnesting the future chain, rewrite it as a coroutine [Avi]
- patch#2: replace `co_await ... finally` with `co_await with_closeable` [Avi]

File: sstables/sstables.cc (hunk 5 above)
Comment (avikivity) [reply sstables/sstables.cc:3215]: Metrics labels. A metric can be tagged with a set of labels and then each unique set of label instances becomes its own time series. If we had a label storage_class=object|local then we could count object storage separately from local storage (and sum them if wanted). We always have implicit labels on the shard and instance.

**Comment by lersek** [comment]
Failed unit tests:
* test_topology_streaming_failure --> #20554 (`raft_topology - raft_state_monitor_fiber aborted with raft::stopped_error (Raft instance is stopped, reason: "raft group0 is stopped")`)
* test_initial_token --> seemingly #20637, but after checking the servers' logs, it's again #20554 in reality

File: sstables/sstables.cc (hunk 5 above)
Comment (lersek) [reply sstables/sstables.cc:3215]: done in v3

**Comment by lersek** [comment]
Posted the v3 series to the mailing list (`[PATCH v3 0/3] sstables: measure storage usage per storage type per shard`); can't provide a direct URL yet because (I assume) as a new list member, I must pass moderation first.
